### ISSUE TITLE
GitHub Adapter/Service Updates

### ISSUE THREAD

**Issue by taylorperkins** [NerdWalletOSS/shepherd#670]
Hey hey! Thank you all for putting this together. Very cool project! I am not a JS/TS dev, per se, so please be patient with me ❤️ However, I am happy to help/contribute if possible. My team at [iheartradio](https://github.com/iheartradio/) is considering a tool like this, and so far sheperd does pretty much everything we would want! However, we noticed a couple of things that may need to be tweaked (at least for our use-case). Could definitely be user-error though!

I started playing around with shepherd yesterday (great tutorial btw!) and noticed a couple of things that were not working out. 

## Context
* iheartradio is a part of the [iheart-media-inc](https://github.com/enterprises/iheart-media-inc/) enterprise
* iheartradio would be using the GitHub service/adapter
* The contents of the `shepherd.yml` is: 
```yaml
id: 2024.01.22-sync-files
title: Sync all watched files
adapter:
  type: github
  search_type: repositories
  search_query: org:iheartradio topic:mle
hooks:
  should_migrate:
    - ls "$SHEPHERD_MIGRATION_DIR/watchers" | grep $(basename "$PWD") # Check that the repo has a watch configuration
  apply: "${SHEPHERD_MIGRATION_DIR}/sync.sh"
  pr_message: echo 'Hey! This PR syncs all files'
```

So - we are trying to apply the `sync.sh` script to all repositories in the `iheartradio` org that contains the `mle` topic. 

I was able to get this working for us using this yaml config, just not at first.

## Checkout Search Issue
When trying out the first step, `checkout`, I was not able to get it to work using the enterprise URL. I believe this errors out when trying to query for repositories based on the `shepherd.yml`. 

For example:
```
export SHEPHERD_GITHUB_ENTERPRISE_URL=api.github.com/enterprises/iheart-media-inc
```
Which produces:
```
RequestError [HttpError]: Not Found
...
{
    url: 'https://api.github.com/enterprises/iheart-media-inc/api/v3/search/repositories?q=org%3Aiheartradio%20topic%3Amle',
    status: 404,
    headers: {
      'access-control-allow-origin': '*',
      'access-control-expose-headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset',
      'content-encoding': 'gzip',
      'content-security-policy': "default-src 'none'",
      'content-type': 'application/json; charset=utf-8',
      date: 'Tue, 23 Jan 2024 19:12:07 GMT',
      'referrer-policy': 'origin-when-cross-origin, strict-origin-when-cross-origin',
      server: 'GitHub.com',
      'strict-transport-security': 'max-age=31536000; includeSubdomains; preload',
      'transfer-encoding': 'chunked',
      vary: 'Accept-Encoding, Accept, X-Requested-With',
      'x-accepted-oauth-scopes': 'repo',
      'x-content-type-options': 'nosniff',
      'x-frame-options': 'deny',
      'x-github-api-version-selected': '2022-11-28',
      'x-github-media-type': 'github.v3; format=json',
      'x-github-request-id': 'D231:63BF:105310:216FB1:65B00F87',
      'x-oauth-scopes': 'admin:org, repo, workflow, write:discussion',
      'x-ratelimit-limit': '5000',
      'x-ratelimit-remaining': '4988',
      'x-ratelimit-reset': '1706038594',
      'x-ratelimit-resource': 'core',
      'x-ratelimit-used': '12',
      'x-xss-protection': '0'
    },
    data: {
      message: 'Not Found',
      documentation_url: 'https://docs.github.com/rest'
    }
  },
  request: {
    method: 'GET',
    url: 'https://api.github.com/enterprises/iheart-media-inc/api/v3/search/repositories?q=org%3Aiheartradio%20topic%3Amle',
    headers: {
      accept: 'application/vnd.github.v3+json',
      'user-agent': 'octokit-rest.js/20.0.2 octokit-core.js/5.1.0 Node.js/19.7.0 (darwin; x64)',
      authorization: 'token [REDACTED]'
    },
    request: { hook: [Function: bound bound register] }
  }
}
```

The `404 NOT FOUND` suggested that the endpoint does not exist for the API, and the permissions for the token I am using look pretty good (`'x-oauth-scopes': 'admin:org, repo, workflow, write:discussion',`). So - I started poking around at the API using Postman to try and recreate this based on the GH API docs. Ultimately, I was able to get this endpoint to work: 
```
https://api.github.com/search/repositories?q=org:iheartradio&topic:mle
```
This returned all the repos in the org; I think `octokit` does some post-response filtering on topic within the code (I could be wrong though).

So - this told me a couple of things:
* The enterprise URL doesn't matter - I can just use the org, `iheartradio`, as part of the query params
* the [`api/v3` suffix](https://github.com/NerdWalletOSS/shepherd/blob/main/src/services/github.ts#L40) that gets appended to the `baseUrl` of the octokit instance is perhaps not needed anymore? 

I cloned down the repository locally and made the change. Then ran:
```
npm run build
npm pack --pack-destination ~
```
Then updated `package.json` to reference my local tar as a dependency. And finally:
```
npm install -g
```
to make the local `sheperd` CLI available globally. After testing.. The search worked 🙌 

## Checkout Clone Issue
Once the above error was resolved, the `Checking out repo` spinning wheel would hang, and ultimately produce an error like: 
```
GitError: Cloning into '$HOME/.shepherd/2024.01.22-sync-files/repos/iheartradio/$REPO'...
ssh: connect to host api.github.com port 22: Operation timed out
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

 task: {
    commands: [
      'clone',
      '--depth',
      '1',
      'git@api.github.com:iheartradio/$REPO.git',
      '$HOME/.shepherd/2024.01.22-sync-files/repos/iheartradio/$REPO'
    ],
    format: 'utf-8',
    parser: [Function: parser]
  }

```
The SSH URL in the error message looked odd, since it contained the `api.` bit. Digging a little more, I noticed that the clone piece was using the [`getRepositoryUrl` method](https://github.com/NerdWalletOSS/shepherd/blob/main/src/adapters/github.ts#L311-L313), which seems like it should be creating an ssh URL, which includes the `api.` bit from the `shepherdGitHubEnterpriseUrl`. Once the `api.` is removed from the string, this works as well. 

 ## Contributing
I have these changes on a branch locally. I'm happy to push up to contribute, but need permissions first 😄 If there is a more in-depth contribution guide to follow, I'm happy to as well. 

Again - thanks for the project! We're really hoping to use this.

**Comment by kavitha186** [comment]
Hi @taylorperkins can you please confirm if the checkout is success without passing topic in the shepherd.yml?

**Comment by taylorperkins** [comment]
Hey @aorinevo and @kavitha186 , thanks for the quick replies! That all makes sense. I'll put up a PR shortly!

RE: setting the `shepherdenterpriseurl` to github.com - Kind of. If I unset the `shepherdenterpriseurl` variable and use the default, [`api.github.com`](https://github.com/NerdWalletOSS/shepherd/blob/main/src/services/github.ts#L13), it works, but only when I remove the `api/v3` from the [baseUrl](https://github.com/NerdWalletOSS/shepherd/blob/main/src/services/github.ts#L40) that gets used to make API calls

**Comment by taylorperkins** [comment]
So - even though iheartradio is a part of the iheart-media-inc enterprise, I don't _need_ to set the `shepherdenterpriseurl`. I am not sure why tbh

**Comment by aorinevo** [comment]
I'm thinking the best way to address this is to have the full enterprise url overwritten using the SHEPHERD_GITHUB_ENTERPRISE_URL env var and default to api.github.com, which would actually align with the env vars name.  That should address your use case @taylorperkins and would also take care of GitHub Enterprise Server vs Cloud use cases.  

Just initial thoughts.  What do you think?

**Comment by taylorperkins** [comment]
Sure! That makes sense. One issue I ran into is that shepherd references two types of URLs:
* One to reference the GitHub API
* One to clone the repository

Does your suggestion handle that as well? I'm happy to test anything out. I _think_ the URL needed to clone the repository should have the same format as the following, but I could be wrong. 
![Screen Shot 2024-01-25 at 9 25 13 AM](https://github.com/NerdWalletOSS/shepherd/assets/24389891/f4c8f389-0c9a-4f66-b810-a2c5b31ba333)

The call made to clone the repository references `SHEPHERD_GITHUB_ENTERPRISE_URL`. If it contains `api.github.com`, then the formats no longer align and it fails. That's what I experienced anyways
